app/scripts/modules/core/src/delivery/service/executions.transformer.service.ts

```diff
@@ -0,0 +1,323 @@
+import { ExecutionBarLabel } from 'core/pipeline/config/stages/core/ExecutionBarLabel';
+import { ExecutionMarkerIcon } from 'core/pipeline/config/stages/core/ExecutionMarkerIcon';
+import { IExecution, IExecutionStage, IExecutionStageSummary, IOrchestratedItem } from 'core/domain';
+import { ORCHESTRATED_ITEM_TRANSFORMER, OrchestratedItemTransformer } from 'core/orchestratedItem/orchestratedItem.transformer';
+import { PIPELINE_CONFIG_PROVIDER, PipelineConfigProvider } from 'core/pipeline/config/pipelineConfigProvider';
+
+export class ExecutionsTransformerService {
+  private hiddenStageTypes = ['initialization', 'pipelineInitialization', 'waitForRequisiteCompletion', 'determineTargetServerGroup'];
+
+  constructor(private orchestratedItemTransformer: OrchestratedItemTransformer,
+              private pipelineConfig: PipelineConfigProvider) {
+    'ngInject';
+  }
+
+  private addDeploymentTargets(execution: IExecution): void {
+    const targets: string[] = [];
+    execution.stages.map(stage => {
+      const stageConfig = this.pipelineConfig.getStageConfig(stage);
+      if (stageConfig && stageConfig.accountExtractor) {
+        targets.push(stageConfig.accountExtractor(stage));
+      }
+    });
+    execution.deploymentTargets = uniq(flattenDeep(targets)).sort();
+  }
+
+  private siblingStageSorter(a: IOrchestratedItem, b: IOrchestratedItem): number {
+    if (!a.startTime && !b.startTime) {
+      return 0;
+    }
+    if (!a.startTime) {
+      return 1;
+    }
+    if (!b.startTime) {
+      return -1;
+    }
+    return a.startTime - b.startTime;
+  }
+
+  private flattenStages(stages: IExecutionStage[], stage: IExecutionStage | IExecutionStageSummary): IExecutionStage[] {
+    if (stage.before && stage.before.length) {
+      stage.before.sort((a, b) => this.siblingStageSorter(a, b));
+      stage.before.forEach((beforeStage) => this.flattenStages(stages, beforeStage));
+    }
+    if ((stage as IExecutionStageSummary).masterStage) {
+      stages.push((stage as IExecutionStageSummary).masterStage);
+    } else {
+      stages.push(stage as IExecutionStage);
+    }
+    if (stage.after && stage.after.length) {
+      stage.after.sort((a, b) => this.siblingStageSorter(a, b));
+      stage.after.forEach((afterStage) => this.flattenStages(stages, afterStage));
+    }
+    return stages;
+  }
+
+  private flattenAndFilter(stage: IExecutionStage | IExecutionStageSummary): IExecutionStage[] {
+    return this.flattenStages([], stage)
+      .filter(s => s.isFailed ||
+        (!this.hiddenStageTypes.includes(s.type) && s.initializationStage !== true));
+  }
+
+  private getCurrentStages(execution: IExecution) {
+    const currentStages = execution.stageSummaries.filter((stage) => stage.isRunning);
+    // if there are no running stages, find the first enqueued stage
+    if (!currentStages.length) {
+      const enqueued = execution.stageSummaries.filter((stage) => stage.hasNotStarted);
+      if (enqueued && enqueued.length) {
```

Comment:
could use `.find` instead of `.filter` above, then skip the `.length` check here